client/telemetry/src/lib.rs

```diff
@@ -333,61 +371,83 @@ impl TelemetryHandle {
 	///
 	/// The `connection_message` argument is a JSON object that is sent every time the connection
 	/// (re-)establishes.
-	pub fn start_telemetry(
-		&mut self,
-		span: TelemetrySpan,
-		endpoints: TelemetryEndpoints,
-		connection_message: ConnectionMessage,
-	) -> TelemetryConnectionNotifier {
-		let Self { message_sender } = self;
+	pub fn start_telemetry(&mut self, connection_message: ConnectionMessage) -> Result<()> {
+		let Self {
+			message_sender: _,
+			register_sender,
+			id,
+			connection_notifier: _,
+			endpoints,
+		} = self;
 
-		let connection_notifier = TelemetryConnectionNotifier {
-			message_sender: message_sender.clone(),
-			addresses: endpoints.0.iter().map(|(addr, _)| addr.clone()).collect(),
+		let endpoints = match endpoints.take() {
+			Some(x) => x,
+			None => return Err(Error::TelemetryAlreadyInitialized),
 		};
 
-		match span.0.id() {
-			Some(id) => {
-				match message_sender.unbounded_send(Register::Telemetry {
-					id,
-					endpoints,
-					connection_message,
-				}) {
-					Ok(()) => {}
-					Err(err) => error!(
-						target: "telemetry",
-						"Could not initialize telemetry: \
-						the telemetry is probably already running: {}",
-						err,
-					),
-				}
-			}
-			None => error!(
-				target: "telemetry",
-				"Could not initialize telemetry: the span could not be entered",
-			),
+		if register_sender
+			.unbounded_send(Register::Telemetry {
+				id: *id,
+				endpoints,
+				connection_message,
+			})
+			.is_err()
+		{
+			return Err(Error::TelemetryWorkerDropped);
 		}
 
-		connection_notifier
+		Ok(())
+	}
+
+	/// Make a new clonable handle to this [`Telemetry`]. This is used for reporting telemetries.
+	pub fn handle(&self) -> TelemetryHandle {
+		TelemetryHandle {
+			message_sender: self.message_sender.clone(),
+			id: self.id,
+			connection_notifier: self.connection_notifier.clone(),
+		}
+	}
+}
+
+/// Handle to a [`Telemetry`]. Used to report telemetries.
+#[derive(Debug, Clone)]
+pub struct TelemetryHandle {
+	message_sender: mpsc::Sender<TelemetryMessage>,
+	id: Id,
+	connection_notifier: TelemetryConnectionNotifier,
+}
+
+impl TelemetryHandle {
+	/// Send telemetries.
+	pub fn send_telemetry(&mut self, verbosity: VerbosityLevel, payload: TelemetryPayload) {
```

Comment:
The problem with making this mut here is that you clone this object very often. Cloning is the same as using an unbounded channel, because every sender is getting at least one spot in the channel always.

I think we should solve this by having some mutex wrapped channel. Maybe some optimization could be to put this into a thread local 🤔

Reply:
Yeah this mut is very annoying tbh... I will check what I can do with the thread local. Sounds interesting!

Reply:
> The problem with making this mut here is that you clone this object very often. Cloning is the same as using an unbounded channel, because every sender is getting at least one spot in the channel always.

I didn't realize that! Thanks for sharing this insight

Reply:
Fixed in fc1ce16

I tried to do the optimization you suggested but it looks... weird. I'm not sure it's any good. Can you check? c1fac06